### PR TITLE
#169817886:user edit trip request 

### PR DESCRIPTION
**Pull request by mujohn26**
#### What does this PR do?

-  user should be able to edit trip request which are still open 

#### Description of Task to be completed?

-  user should provide JWT to be checked
-  user should pass the trip id in the header
-  user should send the editable request on that specific trip
#### How should this be manually tested?
-  clone it 
- run npm i to install dependencies
- run `npm run db-migrate` to create databases
- run `npm  run start`
-  use `pi/v1/trips/:tripId`

#### Any background context you want to provide?

- you must provide JWT token to able to perform the task
- you must be logged in and your account is verified
- you will be able to edit the only request which is pending or rejected 

#### What are the relevant pivotal tracker stories?

[#169817886](https://www.pivotaltracker.com/story/show/169817886)
#### Screenshots (if appropriate)

Screen Shot 2020-01-16 at 21 05 52

#### Questions:

N/A

### CODE REVIEW

File: src/controllers/trip.controller.js [hunk 1]
```diff
@@ -39,27 +39,42 @@ class tripController {
       const status = 'pending';
       const { type } = req.body;
 
-      const createdTrip = await tripService.CreateTrip(req, req.body, tripid, type);
+      const createdTrip = await tripService.CreateTrip(
+        req,
+        req.body,
+        tripid,
+        type
+      );
```
Comment (Jajabenit250) [inline src/controllers/trip.controller.js]: Unnecessary change. I think we might have the previous code we don't need to change it ?

File: src/controllers/trip.controller.js [hunk 2]
```diff
@@ -39,27 +39,42 @@ class tripController {
...
 
-
       const userManager = await userService.findUserManager(userId);
       const { managerId } = userManager[0];
       const requestData = {
-        tripId, userId, managerId, status
+        tripId,
+        userId,
+        managerId,
+        status
```
Comment (Jajabenit250) [inline src/controllers/trip.controller.js]: Unnecessary change. I think we might have the previous code we don't need to change it ?

File: src/controllers/trip.controller.js [hunk 3]
```diff
@@ -39,27 +39,42 @@ class tripController {
...
-      return response.errorMessage(
+      await NotificationService.sendNotification(
+        'trip_request_event',
+        managerId,
+        `New ${type} trip request.`,
+        `${req.user.firstName} has requested a new ${type}`,
+        tripRequest.id,
+        'http://localhost:3000/api/v1/notifications'
+      );
+      return response.successMessage(
```
Comment (Jajabenit250) [inline src/controllers/trip.controller.js]: Unnecessary change. I think we might have the previous code we don't need to change it ?

**Review by Jajabenit250 — COMMENTED** [review]
Good progress so far but kindly attend to my feedback. the general feedback is that you are making many changes and it make reviewer to not know what you actually did.

File: src/controllers/trip.controller.js (hunk 2 above)
Comment (mujohn26) [reply src/controllers/trip.controller.js]: Thanks for the feedback benit, surely I have not made any change about the code and the functionality the reason behind, it's because I used prettier formater to have a good design, on other side no changes I made.

Comment (Jajabenit250) [reply src/controllers/trip.controller.js]: @mujohn26 Kindly check if it is possible to remain with old code that you are not changing. because unnecessary changes will cause merge conflict in future.

File: src/middlewares/accomodation.middleware.js [hunk 4]
```diff
@@ -1,6 +1,7 @@
 import response from '../helpers/response.helper';
-import tripServicres from '../services/trip.services';
-import db from '../database/models';
+// import tripServicres from '../services/trip.services';
+// import db from '../database/models';
```
Comment (bdushimi) [inline src/middlewares/accomodation.middleware.js]: @mujohn26 Kindly always remove commented lines.

File: src/controllers/trip.controller.js [hunk 5]
```diff
@@ -56,7 +56,7 @@ class TripController {
       };
       const tripRequest = await tripService.CreateTripRequest(requestData);
       await NotificationService.sendNotification('trip_request_event', managerId, `New ${type} trip request.`, `${req.user.firstName} has requested a new ${type}`, tripRequest.id, 'http://localhost:3000/api/v1/notifications');
```
Comment (bdushimi) [inline src/controllers/trip.controller.js]: @mujohn26 May you update this URL (http://localhost:3000/api/v1/notifications')) to point to a dynamic URL, that uses the environment variable instead of hardcoding it?

File: src/controllers/trip.controller.js [hunk 6]
```diff
@@ -221,5 +222,82 @@ class TripController {
...
   }
+
+  /**
+   * This controller helps to know the type of trip
+   * sothat it can redirect the request to a tageted controller
+   * @param {Object} req request
+   * @param {*} res response
+   * @returns {Object} return user response
+   */
+  static async combineUpdateTripRequestConctroller(req, res) {
```
Comment (bdushimi) [inline src/controllers/trip.controller.js]: @mujohn26 Kindly update the name of this function. Its description and its name are not matching.

File: src/services/Queries.js [hunk 7]
```diff
@@ -296,5 +317,97 @@ class Queries {
...
+   *
+   * @param {object} table to be updating in
+   * @param {integer} tripId trip id to be found and then replace the status
+   * @returns {object} updated data
+   */
+  static async updateRequestStatus(table, tripId) {
+    try {
+      const updatedRequest = await table.update(
+
+
```
Comment (bdushimi) [inline src/services/Queries.js]: @mujohn26 Kindly remove these white spaces

File: src/controllers/trip.controller.js [hunk 8]
```diff
@@ -221,5 +222,82 @@ class TripController {
...
+   * @param {Object} res The response object
+   * @returns {Object} A user object with selected field
+   */
+  static async updateTripRequest(req, res) {
+    try {
+      const { tripId } = req.params;
+
+
+      await tripRequestService.updateTripRequest(tripId, req.body);
+      await tripRequestService.updateRequestStatus(tripId);
```
Comment (bdushimi) [inline src/controllers/trip.controller.js]: @mujohn26 What's the difference between these two functions (at 253 & 254 lines)? Aren't they doing same thing?

File: src/routes/trip.route.js [hunk 9]
```diff
@@ -151,32 +123,44 @@ router.get('/my-trip-requests', verifyToken.headerToken, verifyUser, tripControl
...
-  requestMiddleware.checkIfBodyIsValid,
-  requestMiddleware.checkIfAlreadyChanged,
-  tripController.updateTripRequestStatus
+  verifyUser,
+  TripMiddleware.checkTripRequestStatus,
+  TripMiddleware.checkIfDateisValid,
+  TripMiddleware.checkLocations,
+  AccomodationMiddleware.findAccommodationByCity,
+  TripMiddleware.multiCityDataValidation,
+  tripController.combineUpdateTripRequestConctroller
```
Comment (bdushimi) [inline src/routes/trip.route.js]: @mujohn26 A couple of things here :
1. No need to add "Controller" to the name of this function because it is already part of the controller class.
2. This function seems to be updating the tripRequest yet the route is about updating a trip information.

File: src/services/Queries.js [hunk 10]
```diff
@@ -296,5 +317,97 @@ class Queries {
...
+      );
+      return updatedRequest;
+    } catch (error) {
+      return error;
+    }
+  }
+
+  /**
+   *
+   * This method will be user to delete the selected trip in multi city
```
Comment (bdushimi) [inline src/services/Queries.js]: @mujohn26 Kindly proofread your description and make sure they're grammatically correct, concise and are conveying the message easily

File: src/services/request.services.js [hunk 11]
```diff
@@ -24,6 +24,87 @@ class TripRequestService {
...
+      return error;
+    }
+  }
+
+  /** Function to update the status to approved or reject
+   *
+   * @param {string} tripId the id of the selected trip
+   * @returns {object} updated data
+   */
+  static async updateRequestStatus(tripId) {
```
Comment (bdushimi) [inline src/services/request.services.js]: @mujohn26 Which request are you referring to here?

File: src/services/Queries.js [hunk 12]
```diff
@@ -296,5 +317,97 @@ class Queries {
...
+   * @param {String} table the name of the table to updated
+   * @param {string} tripId the id of the selected trip
+   * @param {Object} data it gets the data from req.body
+   * @returns {object} updated data
+   */
+  static async deleteMultiCityTripRequest(table, tripId) {
```
Comment (bdushimi) [inline src/services/Queries.js]: @mujohn26 is this function about deleting only a multi-city request or multi-city trip? I think there should be generic functions to delete trip-request and trip given IDs

**Review by bdushimi — CHANGES_REQUESTED** [review]

File: src/controllers/trip.controller.js (hunk 6 above)
Comment (Jajabenit250) [inline src/controllers/trip.controller.js]: what do you mean by combineUpdateTripRequestController i would like if you kindly check the name twice and rename the function according to what it is doing. and since it is a function inside class called TripController i think there no need of calling it controller.

File: src/controllers/trip.controller.js [hunk 13]
```diff
@@ -221,5 +222,82 @@ class TripController {
...
+      return response.errorMessage(
+        res,
+        e.message,
+        500,
+      );
+    }
+  }
+
+  /**
+   * This controller will help a user to edits multiple city request
```
Comment (Jajabenit250) [inline src/controllers/trip.controller.js]: user to edit instead of edits

File: src/middlewares/trip.middleware.js [hunk 14]
```diff
@@ -99,5 +101,26 @@ class TripMiddleware {
 
     return next();
   }
+
+  /**
+   * this function is used to check status of the trip so that it can edited
```
Comment (Jajabenit250) [inline src/middlewares/trip.middleware.js]: it can be edited.

**Review by Jajabenit250 — COMMENTED** [review]
Good Work. But Kindly attend to change requested.

File: src/controllers/trip.controller.js (hunk 8 above)
Comment (mujohn26) [reply src/controllers/trip.controller.js]: I changed them for a better description,one is updating the trip and another one is updating the trip request status.

File: src/services/request.services.js [hunk 15]
```diff
@@ -24,6 +24,87 @@ class TripRequestService {
...
+  /**
+   * This method will provide a service of updating multi city
+   * a trip data in database
+   * @param {Object} body data posted by user
+   * @param { integer } tripId trip id as integer
+   * @param {Object} tripType type of trip
+   * @returns { Object } response data
+   */
+  static async updateMultiCityTripRequest(body, tripId) {
+    const data = {
```
Comment (shemaeric1234) [inline src/services/request.services.js]: the same here  `updateMultiCityTripRequest`  and `updateTrip` methods are doing the same thing, so you can find a way of using a single method.

File: src/services/Queries.js [hunk 16]
```diff
@@ -299,5 +320,95 @@ class Queries {
...
+  }
+
+  /** Function to update the status to approved or reject
+   *
+   * @param {object} table to be updating in
+   * @param {integer} id trip id to be found and then replace the status
+   * @returns {object} updated data
+   */
+  static async updateRequestStatusById(table, id) {
+    try {
```
Comment (shemaeric1234) [inline src/services/Queries.js:384]: Hey, @mujohn26 you did a good job so far, however, there is something you can change.
`updateRequestStatusById`, `updateMultiCityTripRequest` and `updateTrip`  methods are doing the same thing, so you can find a way of using a single method, and think that can be better.

**Review by jo-rdan — APPROVED** [review]
Good work @mujohn26
